graphql_compiler/schema_generation/schema_graph.py

```diff
@@ -625,55 +616,54 @@ def _get_element_properties(self, class_name, non_link_property_definitions):
-        """Return a PropertyDescriptor corresponding to the non-link property definition."""
+    def _get_graphql_type(self, class_name, property_definition):
+        """Return the GraphQLType corresponding to the non-link property definition."""
         name = property_definition['name']
         type_id = property_definition['type']
         linked_class = property_definition.get('linkedClass', None)
         linked_type = property_definition.get('linkedType', None)
-        qualifier = None
-
-        validate_supported_property_type_id(name, type_id)
 
+        graphql_type = None
         if type_id == PROPERTY_TYPE_LINK_ID:
-            raise AssertionError(u'Found a property of type Link on a non-edge class: '
-                                 u'{} {}'.format(name, class_name))
+            raise AssertionError(u'Found a improperly named property of type Link: '
+                                 u'{} {}. Links must be named either "in" or "out"'
+                                 .format(name, class_name))
         elif type_id in COLLECTION_PROPERTY_TYPES:
             if linked_class is not None and linked_type is not None:
                 raise AssertionError(u'Property "{}" unexpectedly has both a linked class and '
                                      u'a linked type: {}'.format(name, property_definition))
             elif linked_type is not None and linked_class is None:
                 # No linked class, must be a linked native OrientDB type.
-                validate_supported_property_type_id(name + ' inner type', linked_type)
-
-                qualifier = linked_type
+                inner_type = get_graphql_scalar_type_or_raise(name + ' inner type', linked_type)
+                graphql_type = GraphQLList(inner_type)
             elif linked_class is not None and linked_type is None:
                 # No linked type, must be a linked non-graph user-defined type.
                 if linked_class not in self._non_graph_class_names:
                     raise AssertionError(u'Property "{}" is declared as the inner type of '
                                          u'an embedded collection, but is not a non-graph class: '
                                          u'{}'.format(name, linked_class))
-
-                qualifier = linked_class
+                if class_name in self._non_graph_class_names:
+                    raise AssertionError('Class {} is a non-graph class that contains a '
+                                         'collection property {}. Only graph classes are allowed '
+                                         'to have collections as properties.'
+                                         .format(class_name, property_definition))
+                # Don't include the fields and implemented interfaces, this information is already
+                # stored in the SchemaGraph.
+                graphql_type = GraphQLList(GraphQLObjectType(linked_class, {}, []))
```

Comment:
It seems like:
- We have multiple, concurrently-existing `GraphQLObjectType` objects that refer to the same type name.
- Some such objects have no field or interface information, making them invalid.

This is a very unusual design decision, so I'd like some more detail on how we intend to make sure this doesn't cause bugs going forward.

Reply:
I wanted to avoid the code complexity involved in properly building the GraphQLObject. (Specially, since it would involve building all of the object's superclasses). To reduce this complexity and deal with the two issues above, how about we limit collection classes to non-Graph classes with no superclasses? I would address this in the next pr.